pkg/cmd/server/start/start_master.go

```diff
@@ -592,31 +592,61 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 		controllerManagerOptions = kc.ControllerManager
 	}
 
-	// Start these first, because they provide credentials for other controllers' clients
-	oc.RunServiceAccountsController()
-	oc.RunServiceAccountTokensController(controllerManagerOptions)
-	// used by admission controllers
-	oc.RunServiceAccountPullSecretsControllers()
-	oc.RunSecurityAllocationController()
-
-	_, _, _, binderClient, err := oc.GetServiceAccountClients(bootstrappolicy.InfraPersistentVolumeBinderControllerServiceAccountName)
-	if err != nil {
-		glog.Fatalf("Could not get client for persistent volume binder controller: %v", err)
+	rootClientBuilder := controller.SimpleControllerClientBuilder{
+		ClientConfig: &oc.PrivilegedLoopbackClientConfig,
 	}
 
-	_, _, _, attachDetachControllerClient, err := oc.GetServiceAccountClients(bootstrappolicy.InfraPersistentVolumeAttachDetachControllerServiceAccountName)
-	if err != nil {
-		glog.Fatalf("Could not get client for attach detach controller: %v", err)
+	// openshiftControllerPreContext represents a context used to start controllers that
+	// requires privileged 'rootClientBuilder' iow. have to start before any other
+	// controller starts (including kubernetes controllers).
+	openshiftControllerContext := origincontrollers.ControllerContext{
+		KubeControllerContext: kctrlmgr.ControllerContext{Options: *controllerManagerOptions},
+		RootClientBuilder:     &rootClientBuilder,
+		ClientBuilder: origincontrollers.OpenshiftControllerClientBuilder{
+			ControllerClientBuilder: controller.SAControllerClientBuilder{
+				ClientConfig:         restclient.AnonymousClientConfig(&oc.PrivilegedLoopbackClientConfig),
+				CoreClient:           oc.PrivilegedLoopbackKubernetesClientsetExternal.Core(),
+				AuthenticationClient: oc.PrivilegedLoopbackKubernetesClientsetExternal.Authentication(),
+				Namespace:            bootstrappolicy.DefaultOpenShiftInfraNamespace,
+			},
+		},
+		DeprecatedOpenshiftInformers: oc.Informers,
+		Stop: utilwait.NeverStop,
 	}
 
-	_, _, _, serviceLoadBalancerClient, err := oc.GetServiceAccountClients(bootstrappolicy.InfraServiceLoadBalancerControllerServiceAccountName)
-	if err != nil {
-		glog.Fatalf("Could not get client for pod gc controller: %v", err)
-	}
+	openshiftControllerPreInitializers, err := oc.NewOpenShiftControllerPreInitializers()
+	allowedOpenshiftControllers := sets.NewString(
+		// The serviceaccount-token controller has to be started before any other controller
+		// runs in order to be able to issue tokens.
+		"serviceaccount-tokens",
```

Comment:
I think we can more or less hard code this one first, then start it directly.  The others simply flow from it.